platform/logger.cpp

```diff
@@ -0,0 +1,252 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(MBED_CONF_RTOS_PRESENT) && !defined(NDEBUG)
+#include <string.h>
+#include "platform/logger_internal.h"
+#include "hal/ticker_api.h"
+#include "hal/us_ticker_api.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_interface.h"
+#include "mbed_events.h"
+
+using namespace mbed;
+
+#if defined (MBED_ID_BASED_TRACING)
+static LOG_DATA_TYPE_ log_id_buf_[LOG_SINGLE_STR_SIZE_];
+static uint32_t log_count_ = 0;
+#endif
+
+// Globals related to ISR logging
+static LOG_DATA_TYPE_ log_isr_str_[LOG_SINGLE_STR_SIZE_];
+static EventQueue *log_isr_equeue_;
+
+// Globals related to time printing
+static const ticker_data_t *const log_ticker_ = get_us_ticker_data();
+static bool time_enable_ = true;
+
+// Globals related to external buffer capturing
+static LOG_DATA_TYPE_ *extern_buf_ = NULL;
+static uint32_t xbuf_count_ = 0;
+
+static void log_update_buf(const char *data, uint32_t size)
+{
+    if ((NULL == extern_buf_) || (NULL == data) || (0 == size)) {
+        return;
+    }
+    uint32_t count = 0;
+    while(size--)
+    {
+        extern_buf_[xbuf_count_++] = data[count++];
+        xbuf_count_ %= MBED_CONF_EXTERNAL_BUFFER_SIZE;
+    }
+}
+
+#if defined (MBED_ID_BASED_TRACING)
+// Note: ISR data is lossy, and with too many interrupts it might be
+// over-written and last valid ISR data will be printed.
+static void log_isr_id_data()
+{
+    mbed_log_lock();
+    uint32_t count = 0;
+    while(log_count_--) {
+        fprintf(stderr, "0x%x ", log_isr_str_[count++]);
+    }
+    mbed_log_unlock();
+}
+#else
+// Note: ISR data is lossy, and with too many interrupts it might be
+// over-written and last valid ISR data will be printed.
+static void log_isr_queue()
+{
+    mbed_log_lock();
+    fputs(log_isr_str_, stderr);
```

Comment:
I'm not okay happy with printing even one line from the high priority background queue - it's too slow - read the performance characteristics of the documentation - it should be basically treated as-if interrupt - ie serial output not acceptable.

Move this to the normal-priority queue.

It would be nice to have at least some protection against ISR tearing and indication of loss - at present this is potentially dangerous as there's no synchronisation of log_isr_str_ at all, so it could happily run off the end of the buffer if reading as it is written.

Reply:
Added protection, but in case of lossy print data from ISR string only option for user at present is to enable buffering to see all ISR logs.